# Used Cars eBay Dataset Clean

This data set is from eBay Kleinanzeigen, a classified section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

Dataquest has made the following changes to the dataset for this project:

* 50,000 data points have been sampled from the original dataset


* the data has been dirtied to resemble a more realistic representation of a scraped data set 

The goal of this project is to clean the data and analyse the used car listings.

In [1]:
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
file_name = 'autos.csv'
autos.to_csv(r'C:\Users\Martin\Desktop\{}'.format(file_name), header=True, index = False)

## Explore The Dataset And Determine Cleaning Tasks

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [6]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [7]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos['offerType'].value_counts()

Angebot    49999
Gesuch         1
Name: offerType, dtype: int64

In [9]:
autos['nrOfPictures'].value_counts()

0    50000
Name: nrOfPictures, dtype: int64

* The dataset has **`50,000`** rows, **`20`** columns, and **`2`** data types (object and int64)


* The columns **`vehicleType`**, **`gearbox`**, **`model`**, **`fuelType`**, and **`notRepairedDamage`** have null values


**Drop Columns** the columns **`seller`** and **`offerType`** have 49999 rows out of 50000 the same value and **`nrOfPictures`** has all 50000 rows the same value so I will drop these columns from the dataset


**Update Column Names** all column names except **`abtest`** and **`gearbox`** are camel case. I will convert all column names to snake case and change the column name **`odometer`** to **`odemeter_km`** to make the units clear 


**Change German Words To English** the columns **`gearbox`**, **`fuelType`**, and **`notRepairedDamage`** contain german words that may be easier to work with if converted to english


**Change Column Type** the columns **`price`** and **`odometer`** are type **`object`**. I will convert these to **`int`** so they can be analysed

## Drop Columns

In [10]:
autos.drop(['seller', 'offerType', 'nrOfPictures'], axis=1, inplace=True)
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicleType  yearOfRegistration    gearbox  powerPS  \
0  $5,000  control         bus                2004    manuell      158   
1  $8,500  control   limousine                1997  automatik      286   
2  $8,990     test   limousine                2009    manuell      102   
3  $4,350  control  kleinwagen                2007  automatik       71   
4  $1,350     test       kombi                2003    manuell        0   

    model   odometer  monthOfRegistration fuelType       brand  \
0  andere  150,000km                    3      lpg     peugeot   
1     7er  150,000km                    6   benzin         bmw   
2    golf   70,000km                    7   benzin  volkswagen   
3  fortwo   70,000km                    6   benzin       smart   
4   focus  150,000km                    7   benzin        ford   

  notRepairedDamage          dateCreated  postalCode             lastSeen  
0              nein  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

## Update Column Names

In [11]:
autos.columns

Index(['dateCrawled', 'name', 'price', 'abtest', 'vehicleType',
       'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer',
       'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage',
       'dateCreated', 'postalCode', 'lastSeen'],
      dtype='object')

In [12]:
autos.columns = ['date_crawled', 'name', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code',
       'last_seen_online']
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gear_box', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen_online'],
      dtype='object')

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model odometer_km  registration_month fuel_type       brand  \
0  andere   150,000km                   3       lpg     peugeot   
1     7er   150,000km                   6    benzin         bmw   
2    golf    70,000km                   7    benzin  volkswagen   
3  fortwo    70,000km                   6    benzin       smart   
4   focus   150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code     last_seen_online  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Change German Words To English

In [14]:
autos['gear_box'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [15]:
gear_box = {'manuell': 'manual',
           'automatik': 'automatic'}
autos['gear_box'] = autos['gear_box'].map(gear_box)
autos['gear_box'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [16]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [17]:
fuel_type = {'lpg':'lpg',
             'benzin':'petrol',
             'diesel':'diesel',
             'cng': 'cng',
             'hybrid':'hybrid',
             'elektro': 'electric',
             'andere': 'other'}
autos['fuel_type'] = autos['fuel_type'].map(fuel_type)
autos['fuel_type'].unique()

array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electric',
       'other'], dtype=object)

In [18]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [19]:
unrepaired_damage = {'nein': 'no',
                     'ja': 'yes'}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)
autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)

## Change Column Type

In [20]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [21]:
autos['odometer_km'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer_km, dtype: object

Convert the __`price`__ and __`kilometer`__ columns to float type

In [22]:
# clean 'price' column and convert to float
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [23]:
# clean 'odometer_km' column and convert to float
autos['odometer_km'] = autos['odometer_km'].str.replace('km', '').str.replace(',', '').astype(int)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [24]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        50000.000000  50000.000000  
mean             5.723360  50813.627300  
std              3.711984  25779.747957  
min              0.000000   1067.000000  
25%              3.000000  30451.000000  
50%              6.000000  49577.000000  
75%              9.000000  71540.000000  
max             12.000000  99998.000000

In [25]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [26]:
autos['price'].unique().shape

(2357,)

In [27]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [28]:
autos['price'].value_counts().sort_index().head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64

In [29]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [30]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [31]:
autos['odometer_km'].unique().shape

(13,)

In [32]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [33]:
autos = autos[autos['price'].between(500, 100000)]
autos['price'].describe()

count    45058.000000
mean      6177.407342
std       7636.932471
min        500.000000
25%       1500.000000
50%       3500.000000
75%       7900.000000
max      99900.000000
Name: price, dtype: float64

In [34]:
len(autos)

45058

The **`price`** column shows some unrealistic prices, so I will only use values between \\$500 and \\$100,000 dollars. The dataset now has 45058 rows.

## Exploring Date Columns

In [35]:
autos[['date_crawled', 'ad_created', 'last_seen_online']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45058 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   date_crawled      45058 non-null  object
 1   ad_created        45058 non-null  object
 2   last_seen_online  45058 non-null  object
dtypes: object(3)
memory usage: 1.4+ MB


In [36]:
autos[['date_crawled','ad_created','last_seen_online']][0:5]

date_crawled           ad_created     last_seen_online
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The **`date_crawled`**, **`ad_created`**, and **`last_seen_online`** columns are type **`object`**

In [37]:
date_crawled_dist = autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
date_crawled_dist

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64

* The first date crawled was the **`5th March 2016`**

* The last date crawled was the **`7th April 2016`**

* The date crawled is evenly distributed with approximately **`3%`** per day over approximately one month

In [38]:
ad_created_dist = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
ad_created_dist

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, Length: 76, dtype: float64

In [39]:
ad_created_dist.tail()

2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, dtype: float64

In [40]:
ad_created_dist[20:70]

2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
2016-02-22    0.000022
2016-02-23    0.000089
2016-02-24    0.000044
2016-02-25    0.000067
2016-02-26    0.000044
2016-02-27    0.000133
2016-02-28    0.000222
2016-02-29    0.000178
2016-03-01    0.000089
2016-03-02    0.000111
2016-03-03    0.000843
2016-03-04    0.001509
2016-03-05    0.023104
2016-03-06    0.015402
2016-03-07    0.034933
2016-03-08    0.033091
2016-03-09    0.033024
2016-03-10    0.032469
2016-03-11    0.033357
2016-03-12    0.037108
2016-03-13    0.016978
2016-03-14    0.034911
2016-03-15    0.033779
2016-03-16    0.029850
2016-03-17    0.030805
2016-03-18    0.013494
2016-03-19    0.033646
2016-03-20    0.038217
2016-03-21    0.037951
2016-03-22    0.032691
2016-03-23    0.032225
2016-03-24    0.028963
2016-03-25    0.031249
2016-03-26 

In [41]:
ad_created_dist.describe()

count    76.000000
mean      0.013158
std       0.015960
min       0.000022
25%       0.000022
50%       0.000155
75%       0.032508
max       0.039039
Name: ad_created, dtype: float64

* The first ad was created on the **`11th June 2015`**

* The last ad was created on the **`7th April 2016`**

* The ads created column is left skewed with most ads being created in the second half of the approximately 11 month period

In [42]:
last_seen_online_dist = autos['last_seen_online'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
last_seen_online_dist

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen_online, dtype: float64

* The first date an ad was last seen online was the **`5th March 2016`**

* The last date an ad was last seen online was the **`7th April 2016`**

* The last seen online date's distribution is left skewed. Almost 50% of the ads were last seen on the last 3 days of the 34 day period

## Exploring Registration Year

In [43]:
autos['registration_year'].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [44]:
count1 = 0
count2 = 0
count3 = 0

for year in autos['registration_year']:
    if year < 1900:
        count1 += 1
    if year < 1925:
        count2 += 1
    if year < 1950:
        count3 += 1
        
print('num of ads with cars having a registration year before 1900:', count1)
print('num of ads with cars having a registration year before 1925:', count2)
print('num of ads with cars having a registration year before 1950:', count3)

num of ads with cars having a registration year before 1900: 2
num of ads with cars having a registration year before 1925: 4
num of ads with cars having a registration year before 1950: 19


In [45]:
for year in autos['registration_year']:
    if year < 1900:
        print(year)

1000
1001


Since the car was invented around the 1880's (this [website](https://www.daimler.com/company/tradition/company-history/1885-1886.html) says 1886 by Carl Benz) these two listings must be incorrect so I will remove them. Any car with a registration later than 2016 must also be incorrect because the latest ad listing was posted 2016 so I will remove those.

In [46]:
count = 0
for year in autos['registration_year']:
    if year > 2016:
        count += 1
print('num of ads with a registration year later than 2016:', count)

num of ads with a registration year later than 2016: 1772


In [47]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].describe()

count    43284.000000
mean      2003.225349
std          7.065511
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [48]:
autos['registration_year'].value_counts(normalize=True).sort_index().tail(30)

1987    0.001571
1988    0.002957
1989    0.003581
1990    0.006330
1991    0.006769
1992    0.007023
1993    0.007855
1994    0.011552
1995    0.019846
1996    0.024767
1997    0.034909
1998    0.046738
1999    0.059421
2000    0.062702
2001    0.058197
2002    0.055887
2003    0.061385
2004    0.061986
2005    0.066168
2006    0.061593
2007    0.052421
2008    0.050920
2009    0.048009
2010    0.036549
2011    0.037335
2012    0.030127
2013    0.018321
2014    0.015017
2015    0.008317
2016    0.021809
Name: registration_year, dtype: float64

The majority of cars are between **`1999`** and **`2008`**

## Exploring Brand

In [49]:
autos['brand'].value_counts()

volkswagen        9180
bmw               4980
mercedes_benz     4427
opel              4347
audi              3920
ford              2822
renault           1902
peugeot           1272
fiat              1019
seat               775
skoda              743
smart              655
mazda              650
nissan             650
citroen            613
toyota             581
hyundai            440
sonstige_autos     415
mini               406
volvo              406
honda              338
mitsubishi         335
kia                318
alfa_romeo         287
chevrolet          261
suzuki             256
porsche            252
chrysler           156
dacia              123
jeep               106
land_rover          98
daihatsu            94
subaru              86
saab                72
jaguar              70
daewoo              58
rover               54
trabant             48
lancia              43
lada                26
Name: brand, dtype: int64

In [50]:
autos['brand'].describe()

count          43284
unique            40
top       volkswagen
freq            9180
Name: brand, dtype: object

In [51]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212088
bmw               0.115054
mercedes_benz     0.102278
opel              0.100430
audi              0.090565
ford              0.065197
renault           0.043942
peugeot           0.029387
fiat              0.023542
seat              0.017905
skoda             0.017166
smart             0.015133
mazda             0.015017
nissan            0.015017
citroen           0.014162
toyota            0.013423
hyundai           0.010165
sonstige_autos    0.009588
mini              0.009380
volvo             0.009380
honda             0.007809
mitsubishi        0.007740
kia               0.007347
alfa_romeo        0.006631
chevrolet         0.006030
suzuki            0.005914
porsche           0.005822
chrysler          0.003604
dacia             0.002842
jeep              0.002449
land_rover        0.002264
daihatsu          0.002172
subaru            0.001987
saab              0.001663
jaguar            0.001617
daewoo            0.001340
rover             0.001248
t

In [52]:
sum(autos['brand'].value_counts()[:20])

40203

In [53]:
brands_top_20 = autos['brand'].value_counts().index[:20]
brands_top_20

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'smart', 'mazda', 'nissan',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'mini', 'volvo'],
      dtype='object')

The analysis will include the top 20 out of 40 brands which includes 40203 rows of the original 43284

In [54]:
brand_mean_price = {}

for brand in brands_top_20:
    brand_mean_price[brand] = int(autos[autos['brand'] == brand]['price'].mean())

brand_mean_price

{'volkswagen': 5783,
 'bmw': 8447,
 'mercedes_benz': 8666,
 'opel': 3394,
 'audi': 9571,
 'ford': 4247,
 'renault': 2819,
 'peugeot': 3360,
 'fiat': 3256,
 'seat': 4810,
 'skoda': 6558,
 'smart': 3611,
 'mazda': 4459,
 'nissan': 5175,
 'citroen': 4012,
 'toyota': 5266,
 'hyundai': 5686,
 'sonstige_autos': 11509,
 'mini': 10691,
 'volvo': 5189}

In [55]:
bmp_df = pd.DataFrame.from_dict(brand_mean_price, orient='index', columns = ['mean_price'])
bmp_df['mean_price'].sort_values(ascending=False)

sonstige_autos    11509
mini              10691
audi               9571
mercedes_benz      8666
bmw                8447
skoda              6558
volkswagen         5783
hyundai            5686
toyota             5266
volvo              5189
nissan             5175
seat               4810
mazda              4459
ford               4247
citroen            4012
smart              3611
opel               3394
peugeot            3360
fiat               3256
renault            2819
Name: mean_price, dtype: int64

Since **`sonstige`** is German for **`other`**, we'll take that out of the dataframe

In [56]:
bmp_df = bmp_df.drop(index='sonstige_autos')
len(bmp_df)

19

In [57]:
bmp_df['mean_price'].sort_values(ascending=False)

mini             10691
audi              9571
mercedes_benz     8666
bmw               8447
skoda             6558
volkswagen        5783
hyundai           5686
toyota            5266
volvo             5189
nissan            5175
seat              4810
mazda             4459
ford              4247
citroen           4012
smart             3611
opel              3394
peugeot           3360
fiat              3256
renault           2819
Name: mean_price, dtype: int64

* **`mini`** has the **`highest mean price`**

* **`renault`** has the **`lowest mean price`**

* The top 4 brands **`mini`**, **`audi`**, **`mercedes_benz`**, and **`bmw`** stand out from the others as having the highest mean prices

* The bottom 5 brands **`smart`**, **`opel`**, **`peugeot`**, **`fiat`** and **`renault`** stand out as having the lowest mean prices

In [58]:
brands_top_4 = ['mini', 'audi', 'mercedes_benz', 'bmw']
brands_top_4

['mini', 'audi', 'mercedes_benz', 'bmw']

In [59]:
mean_price = {}
mean_mileage = {}

for brand in brands_top_4:
    mean_price[brand] = autos[autos['brand'] == brand]['price'].mean()
    mean_mileage[brand] = autos[autos['brand'] == brand]['odometer_km'].mean()

In [60]:
bmp_series = pd.Series(mean_price)
bmm_series = pd.Series(mean_mileage)

In [61]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df['mean_mileage'] = bmm_series
df

mean_price   mean_mileage
mini           10691.061576   88682.266010
audi            9571.457398  128941.326531
mercedes_benz   8666.677208  131083.126271
bmw             8447.069880  132928.714859

In [62]:
df['mean_price'].corr(df['mean_mileage'])

-0.91162226869708

There is a strong negative correlation between mean price and mean mileage in the top 4 most expensive car brands by mean price. As mean price increases, mean mileage decreases. 